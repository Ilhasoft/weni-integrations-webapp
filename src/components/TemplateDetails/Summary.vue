<template>
  <div class="whatsapp_template_summary">
    <div class="whatsapp_template_summary__title">
      <div class="whatsapp_template_summary__title__main">
        {{ $t('WhatsApp.template_details.summary.performace_overview') }}
      </div>
      <div class="whatsapp_template_summary__title__subtitle">
        {{ $t('WhatsApp.template_details.summary.last_7_days') }}
      </div>
    </div>
    <div class="whatsapp_template_summary__main">
      <div class="whatsapp_template_summary__main__card">
        <div class="whatsapp_template_summary__main__card__title">
          {{ $t('WhatsApp.template_details.summary.sent_messages') }}
          <unnnic-icon icon="info" size="sm" filled scheme="neutral-cleanest" />
        </div>
        <div class="whatsapp_template_summary__main__card__value">{{ weekValues.sent }}</div>
      </div>
      <div class="whatsapp_template_summary__main__card">
        <div class="whatsapp_template_summary__main__card__title">
          {{ $t('WhatsApp.template_details.summary.delivered_messages') }}
          <unnnic-icon icon="info" size="sm" filled scheme="neutral-cleanest" />
        </div>
        <div class="whatsapp_template_summary__main__card__value">{{ weekValues.delivered }}</div>
      </div>
      <div class="whatsapp_template_summary__main__card">
        <div class="whatsapp_template_summary__main__card__title">
          {{ $t('WhatsApp.template_details.summary.read_messages') }}
          <unnnic-icon icon="info" size="sm" filled scheme="neutral-cleanest" />
        </div>
        <div class="whatsapp_template_summary__main__card__value">
          {{ weekValues.read }} ({{
            Math.floor((weekValues.read / weekValues.sent) * 100) || '--'
          }}%)
        </div>
      </div>
    </div>
    <div class="whatsapp_template_summary__quality">
      <div class="whatsapp_template_summary__quality__title">
        <div class="whatsapp_template_summary__title__main">
          {{ $t('WhatsApp.template_details.summary.quality_overview') }}
        </div>
        <div class="whatsapp_template_summary__title__subtitle">
          {{ $t('WhatsApp.template_details.summary.last_7_days') }}
        </div>
      </div>
      <div class="whatsapp_template_summary__quality__status">
        <div class="whatsapp_template_summary__quality__status__level">
          <div class="whatsapp_template_summary__quality__status__level__title">
            {{ $t('WhatsApp.template_details.summary.current_level') }}
          </div>
          <div class="whatsapp_template_summary__quality__status__level__value">--</div>
        </div>
        <div class="whatsapp_template_summary__quality__status__blocked">
          <div class="whatsapp_template_summary__quality__status__blocked__title">
            {{ $t('WhatsApp.template_details.summary.block_reason') }}
          </div>
          <div class="whatsapp_template_summary__quality__status__blocked__value">--</div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
  import { mapState, mapActions } from 'vuex';
  export default {
    name: 'Summary',
    mounted() {
      this.fetchTemplateAnalyticsWeek();
    },
    data() {
      return {
        sevenDays: 7 * 24 * 60 * 60 * 1000,
      };
    },
    computed: {
      ...mapState({
        templateAnalytics: (state) => state.insights.templateAnalytics,
        appUuid: (state) => state.insights.appUuid,
        selectedTemplate: (state) => state.insights.selectedTemplate,
      }),
      weekValues() {
        return (
          this.templateAnalytics?.grand_totals || {
            sent: 0,
            delivered: 0,
            read: 0,
          }
        );
      },
    },
    methods: {
      ...mapActions('insights', ['getTemplateAnalytics', 'getTemplates']),
      fetchTemplateAnalyticsWeek() {
        const params = {
          app_uuid: this.appUuid,
          filters: {
            start: this.formatDate(new Date(Date.now() - this.sevenDays)),
            end: this.formatDate(new Date()),
            fba_template_ids: this.selectedTemplate.translations.map(
              (item) => item.message_template_id,
            ),
          },
        };
        this.getTemplateAnalytics(params);
      },
      formatDate(date) {
        return `${date.getMonth() + 1}-${date.getDate()}-${date.getFullYear()}`;
      },
    },
  };
</script>

<style lang="scss">
  @import './styles.scss';
</style>
