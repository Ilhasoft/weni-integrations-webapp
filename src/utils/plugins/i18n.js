import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

import en from '../../locales/en.json';
import pt_br from '../../locales/pt_br.json';
import es_es from '../../locales/es_es.json';

const languages = {
  'en-us': en,
  'pt-br': pt_br,
  es: es_es,
};

const messages = Object.assign(languages);

const i18n = new VueI18n({
  locale: 'en-us',
  fallbackLocale: 'en-us',
  messages,
});

export default i18n;
