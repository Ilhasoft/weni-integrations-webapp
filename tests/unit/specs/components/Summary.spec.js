import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Summary from '@/components/TemplateDetails/Summary.vue';
import i18n from '@/utils/plugins/i18n';
import { selectedTemplate, templateAnalytics, templates } from '../../../__mocks__/appMock';

const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock('@/api/insights', () => {
  return {
    get_template_analytics: jest.fn(),
    get_templates: jest.fn(),
  };
});

describe('components/TemplateDetails/Summary.vue', () => {
  let wrapper;
  let actions;
  let state;
  let store;
  beforeEach(() => {
    actions = {
      getTemplates: jest.fn(),
      getTemplateAnalytics: jest.fn(() => {
        return {
          data: [
            {
              template_id: '730081812069736',
              template_name: null,
              totals: { sent: 51, delivered: 51, read: 30 },
              dates: [
                { start: '2024-01-04', sent: 11, delivered: 10, read: 6 },
                { start: '2024-01-05', sent: 7, delivered: 7, read: 5 },
                { start: '2024-01-06', sent: 6, delivered: 5, read: 1 },
                { start: '2024-01-07', sent: 4, delivered: 5, read: 4 },
                { start: '2024-01-08', sent: 1, delivered: 2, read: 2 },
                { start: '2024-01-09', sent: 0, delivered: 0, read: 0 },
                { start: '2024-01-10', sent: 11, delivered: 11, read: 5 },
                { start: '2024-01-11', sent: 11, delivered: 11, read: 7 },
              ],
            },
          ],
          grand_totals: { sent: 51, delivered: 51, read: 30 },
        };
      }),
    };

    state = {
      insights: {
        isActive: true,
        templateAnalytics: templateAnalytics,
        loadingTemplateAnalytics: false,
        errorTemplateAnalytics: null,
        selectedTemplate: selectedTemplate,
        appUuid: '8e876af8-a59d-4eef-aeb4-61689d2d382b',
        templates: templates,
        errorTemplates: null,
      },
    };
    store = new Vuex.Store({
      actions,
      state,
    });

    wrapper = mount(Summary, {
      localVue,
      i18n,
      store,
      mocks: {
        $t: () => 'some specific text',
      },
    });
  });

  it('should be rendered properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should set weekValues', () => {
    expect(wrapper.vm.weekValues).toEqual(templateAnalytics.grand_totals);
    wrapper.vm.templateAnalytics.grand_totals = null;
    expect(wrapper.vm.weekValues).toEqual({
      sent: 0,
      delivered: 0,
      read: 0,
    });
  });
  it('should call fetchTemplateAnalyticsWeek', () => {
    const spy = spyOn(wrapper.vm, 'fetchTemplateAnalyticsWeek');
    expect(spy).not.toHaveBeenCalled();
    wrapper.vm.fetchTemplateAnalyticsWeek();
    expect(spy).toHaveBeenCalledTimes(1);
  });
  it('should call getTemplateAnalytics()', () => {
    const spy = spyOn(wrapper.vm, 'getTemplateAnalytics');
    expect(spy).not.toHaveBeenCalled();
    wrapper.vm.fetchTemplateAnalyticsWeek();
    expect(spy).toHaveBeenCalledTimes(1);
  });
  it('should call formatDate()', () => {
    const spy = spyOn(wrapper.vm, 'formatDate');
    expect(spy).not.toHaveBeenCalled();
    wrapper.vm.formatDate(new Date());
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
