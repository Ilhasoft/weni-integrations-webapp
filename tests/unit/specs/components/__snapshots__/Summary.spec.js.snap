// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TemplateDetails/Summary.vue should be rendered properly 1`] = `
<div class="whatsapp_template_summary">
  <div class="whatsapp_template_summary__title">
    <div class="whatsapp_template_summary__title__main">
      some specific text
    </div>
    <div class="whatsapp_template_summary__title__subtitle">
      some specific text
    </div>
  </div>
  <div class="whatsapp_template_summary__main">
    <div class="whatsapp_template_summary__main__card">
      <div class="whatsapp_template_summary__main__card__title">
        some specific text
        <unnnic-icon icon="info" size="sm" filled="" scheme="neutral-cleanest"></unnnic-icon>
      </div>
      <div class="whatsapp_template_summary__main__card__value">47</div>
    </div>
    <div class="whatsapp_template_summary__main__card">
      <div class="whatsapp_template_summary__main__card__title">
        some specific text
        <unnnic-icon icon="info" size="sm" filled="" scheme="neutral-cleanest"></unnnic-icon>
      </div>
      <div class="whatsapp_template_summary__main__card__value">48</div>
    </div>
    <div class="whatsapp_template_summary__main__card">
      <div class="whatsapp_template_summary__main__card__title">
        some specific text
        <unnnic-icon icon="info" size="sm" filled="" scheme="neutral-cleanest"></unnnic-icon>
      </div>
      <div class="whatsapp_template_summary__main__card__value">
        28 (59%)
      </div>
    </div>
  </div>
  <div class="whatsapp_template_summary__quality">
    <div class="whatsapp_template_summary__quality__title">
      <div class="whatsapp_template_summary__title__main">
        some specific text
      </div>
      <div class="whatsapp_template_summary__title__subtitle">
        Ãšltimos dias
      </div>
    </div>
    <div class="whatsapp_template_summary__quality__status">
      <div class="whatsapp_template_summary__quality__status__level">
        <div class="whatsapp_template_summary__quality__status__level__title">
          some specific text
        </div>
        <div class="whatsapp_template_summary__quality__status__level__value">--</div>
      </div>
      <div class="whatsapp_template_summary__quality__status__blocked">
        <div class="whatsapp_template_summary__quality__status__blocked__title">
          some specific text
        </div>
        <div class="whatsapp_template_summary__quality__status__blocked__value">--</div>
      </div>
    </div>
  </div>
</div>
`;
